### PR TITLE
Use appropriate integer sizes for query parameters

### LINKED ISSUE #16: All integer named parameters mapped to bigint
**Describe the bug**
Query like:
```go
res, err := db.Query(`SELECT COUNT(*) FROM test WHERE partition_id=@partitionId`,
	sql.Named("partitionId", int(1)))
```
executed as:
```SQL
exec sp_executesql N'SELECT * FROM test WHERE partition_id=@partitionId',N'@partitionId bigint',@partitionId=1
```
`partitionId` parameter defined as `N'@partitionId bigint'` even in code we are using `int` not `int64`

It's important, because plan will have implicit cast and as a result may be suboptimal plan especially when using partitions.

For example, if table was defined with partitions schema (notice that partition column is int):
```SQL
DROP TABLE IF EXISTS test;
			IF EXISTS(SELECT * FROM sys.partition_schemes WHERE name='ps')
			BEGIN
				DROP PARTITION SCHEME ps;
			END
			IF EXISTS(SELECT * FROM sys.partition_functions WHERE name='pf')
			BEGIN			
				DROP PARTITION FUNCTION pf;
			END
			CREATE PARTITION FUNCTION pf (int) AS RANGE LEFT FOR VALUES (0,1,2);
			CREATE PARTITION SCHEME ps AS PARTITION pf ALL TO('PRIMARY');
			CREATE TABLE test(partition_id int, id bigint) ON ps(partition_id);
			CREATE UNIQUE CLUSTERED INDEX pk on test(partition_id, id)
			INSERT INTO test values(1,1);
```
Plan with `,N'@partitionId bigint'` will look like (notice greater/less then comparison):
image

Plan with `,N'@partitionId int'` will look like (equality comparison, expected):
imageNote: problem with partitions is only one case of problems caused by implicit cast.

**To Reproduce**
Run SQL profile to see what is executed: 
```go
package main

import (
	"context"
	"database/sql"
	"fmt"

	"github.com/jmoiron/sqlx"
	_ "github.com/microsoft/go-mssqldb"
)

const CONNECTION_STR = "sqlserver://sa:<yourpassword here>@127.0.0.1:1433?database=D1"

func main() {

			IF EXISTS(SELECT * FROM sys.partition_schemes WHERE name='ps')
			BEGIN
				DROP PARTITION SCHEME ps;
			END
			IF EXISTS(SELECT * FROM sys.partition_functions WHERE name='pf')
			BEGIN			
				DROP PARTITION FUNCTION pf;
			END
			CREATE PARTITION FUNCTION pf (int) AS RANGE LEFT FOR VALUES (0,1,2);
			CREATE PARTITION SCHEME ps AS PARTITION pf ALL TO('PRIMARY');
			CREATE TABLE test(partition_id int, id bigint) ON ps(partition_id);
			CREATE UNIQUE CLUSTERED INDEX pk on test(partition_id, id)
			INSERT INTO test values(1,1);
			`); err != nil {
		fmt.Println("Error", err)
		return
	}

	res, err := db.Query(`SELECT COUNT(*) FROM test WHERE partition_id=@partitionId`,
		sql.Named("partitionId", int(1)))

	if err != nil {
		fmt.Println("Error", err)
		return
	}

	res.Next()
	cnt := 0
	err = res.Scan(&cnt)
	if err != nil {
		fmt.Println("Error", err)
		return
	}

	fmt.Println(cnt)

}
```
In SQL profiler you could see what was executed. Run query in SSMS to see the plan

**Expected behavior**
It should be executed as
```SQL
exec sp_executesql N'SELECT * FROM test WHERE partition_id=@partitionId',N'@partitionId int',@partitionId=1
```
Parameter passed as `,N'@partitionId int'`

### PR DESCRIPTION
**Pull request by shueybubbles**
Fixes #16 
I'd like community input on #21. Changing the scan-to-interface{} behavior is fairly easy to do but it may have undesired side effects if apps are expecting `int64` in that scenario and could fail otherwise.

### CODE REVIEW

File: mssql_go110_perf_test.go [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+	}
+	conn := sql.OpenDB(connector)
+	defer conn.Close()
+	conn.SetMaxOpenConns(1)
+	conn.SetMaxIdleConns(1)
+	rows, err := conn.Query("select 'prime the pump'")
+	if err != nil {
+		b.Fatal("Unable to query")
+	}
+	rows.Close()
```
[the remark this answers is not in the document]
Comment (shueybubbles) [reply mssql_go110_perf_test.go:29]: by design

[the remark this answers is not in the document]
Comment (shueybubbles) [reply mssql_go110_perf_test.go:29]: by design

File: mssql.go [hunk 2]
```diff
@@ -906,6 +907,35 @@ func (s *Stmt) makeParam(val driver.Value) (res param, err error) {
 		return
 	}
 	switch val := val.(type) {
+	case int:
+		res.ti.TypeId = typeIntN
+		// Rather than guess if the caller intends to pass a 32bit int from a 64bit app based on the
```
Comment (kenvanhyning) [inline mssql.go:912]: Is this going to result in a potential breaking change to existing clients?

File: bulkcopy.go [hunk 3]
```diff
@@ -397,6 +405,14 @@ func (b *Bulk) makeParam(val DataValue, col columnStruct) (res param, err error)
...
 		case int64:
 			res.buffer = []byte(strconv.FormatInt(val, 10))
+		case int:
+			res.buffer = []byte(strconv.FormatInt(int64(val), 10))
+		case int8:
+			res.buffer = []byte(strconv.FormatInt(int64(val), 10))
+		case int32:
+			res.buffer = []byte(strconv.FormatInt(int64(val), 10))
+		case int16:
+			res.buffer = []byte(strconv.FormatInt(int64(val), 10))
```
Comment (stuartpa) [inline bulkcopy.go:411]: Can these case statements be ordered int, int8, int16, int32, int64?

File: bulkcopy_test.go [hunk 4]
```diff
@@ -81,6 +82,10 @@ func TestBulkcopy(t *testing.T) {
 		{"test_binary", []byte("1"), nil},
 		{"test_binary_16", bin, nil},
 		{"test_intvarchar", 1234, "1234"},
+		{"test_int64nvarchar", int64(123456), "123456"},
+		{"test_int32nvarchar", int32(12345), "12345"},
+		{"test_int16nvarchar", int16(1234), "1234"},
```
Comment (stuartpa) [inline bulkcopy_test.go:87]: Would it be better if these test values are values that don't fit in the domain of the next smallest type?

File: mssql.go [hunk 5]
```diff
@@ -906,6 +907,35 @@ func (s *Stmt) makeParam(val driver.Value) (res param, err error) {
...
+		// value of the int, we'll just match the runtime size of int.
+		// Apps that want a 32bit int should use int32
+		if bits.UintSize == 32 {
+			res.buffer = make([]byte, 4)
+			res.ti.Size = 4
+			binary.LittleEndian.PutUint32(res.buffer, uint32(val))
+		} else {
+			res.ti.TypeId = typeIntN
```
Comment (stuartpa) [inline mssql.go]: Is this line of code redundant?

File: mssql_go19.go [hunk 6]
```diff
@@ -1,3 +1,4 @@
+//go:build go1.9
```
Comment (stuartpa) [inline mssql_go19.go:1]: Is something adding this // comment?

**Review by stuartpa — APPROVED** [review]
:shipit:

File: mssql.go (hunk 2 above)
Comment (shueybubbles) [reply mssql.go:912]: yeah for any type != Int64 , we will now pass a matching size int instead of a BigInt.

File: mssql.go (hunk 5 above)
Comment (shueybubbles) [reply mssql.go]: yes

File: bulkcopy_test.go (hunk 4 above)
Comment (shueybubbles) [reply bulkcopy_test.go:87]: no. this test just covers the conversion to nvarchar parameter.

File: bulkcopy.go (hunk 3 above)
Comment (shueybubbles) [reply bulkcopy.go:411]: sure

File: mssql.go (hunk 2 above)
Comment (shueybubbles) [reply mssql.go:912]: it's not "breaking" in the sense of an interface change, but it means that possibly some type conversions in existing queries will change. Maybe some apps have TSQL queries that expect all integer parameters to be bigint and now the SQL runtime will have to coerce them on the server side.

**Pull request merged**